harp_apps/proxy/models/remotes.py

```diff
@@ -0,0 +1,254 @@
+        if reason:
+            self.failure_reasons.add(reason)
+
+    def _asdict(self, /, *, secure=True):
+        return {
+            "url": self.url,
+            **(
+                {"failure_threshold": self.failure_threshold}
+                if self.failure_threshold != type(self).failure_threshold
+                else {}
+            ),
+            **(
+                {"success_threshold": self.success_threshold}
+                if self.success_threshold != type(self).success_threshold
+                else {}
+            ),
+        }
+
+
+class HttpProbe:
+    """
+    A ``HttpProbe`` is a health check that can be used to check the health of a remote's endpoints. It is used as the
+    configuration parser for ``proxy.endpoints[].remote.probe`` settings.
+
+    .. code-block:: yaml
+
+        type: http
+        method: GET
+        path: /health
+        headers:
+          x-purpose: "health probe"
+        timeout: 5.0
+    """
+
+    timeout = 10.0
+
+    def __init__(
+        self,
+        method: str,
+        path: str,
+        headers=None,
+        timeout=None,
+        type="http",
+    ):
+        if type != "http":
+            raise ValueError(f"Invalid probe type: {type}")
+
+        self.method = method
+        self.path = path
+        self.headers = headers or {}
+        self.timeout = float(timeout) if timeout else self.timeout
+
+    async def check(self, client: httpx.AsyncClient, url: HttpEndpoint):
+        try:
+            response = await client.request(
+                self.method, urljoin(url.url, self.path), headers=copy(self.headers), timeout=self.timeout
+            )
+            if 200 <= response.status_code < 400:
+                url.probe_success()
+            else:
+                url.probe_failure(f"HTTP_{response.status_code}")
+        except Exception as exc:
+            url.probe_failure(shouty_snake(type(exc).__name__))
+
+    def _asdict(self, /, *, secure=True):
+        return {
+            "type": "http",
+            "method": self.method,
+            "path": self.path,
+            **({"headers": self.headers} if self.headers else {}),
+            **({"timeout": self.timeout} if self.timeout != type(self).timeout else {}),
+        }
+
+
+class HttpRemote:
+    """
+    A ``HttpRemote`` is a collection of endpoints that a proxy will use to route requests. It is used as the
+    configuration parser for ``proxy.endpoints[].remote`` settings.
+
+    .. code-block:: yaml
+
+        min_pool_size: 1
+        endpoints:
+          # see HttpEndpoint
+          - ...
+        probe:
+          # see HttpProbe
+          ...
+
+    """
+
+    endpoints = None
+    current_pool: deque[HttpEndpoint] = None
+    current_pool_name = None
+    min_pool_size = 1
+
+    def __init__(
+        self,
+        endpoints: Iterable[HttpEndpoint | dict | str] | str = (),
+        *,
+        min_pool_size=None,
+        probe: HttpProbe | dict = None,
```

Comment:
linter is complaining about types here

```suggestion
        probe: Optional[HttpProbe | dict] = None,
```